xapian-core/weight/pl2plusweight.cc

```diff
@@ -0,0 +1,153 @@
+/** @file pl2plusweight.cc
+ * @brief Xapian::PL2PlusWeight class - the PL2+ weighting scheme of the DFR framework.
+ */
+/* Copyright (C) 2016 Vivek Pal
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU General Public License as
+ * published by the Free Software Foundation; either version 2 of the
+ * License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA
+ */
+
+#include <config.h>
+
+#include "xapian/weight.h"
+#include "common/log2.h"
+
+#include "serialise-double.h"
+
+#include "xapian/error.h"
+
+using namespace std;
+
+namespace Xapian {
+
+PL2PlusWeight::PL2PlusWeight(double c, double delta)
+	: param_c(c), param_delta(delta)
+{
+    if (param_c <= 0)
+	throw Xapian::InvalidArgumentError("Parameter c is invalid.");
+    if (param_delta <= 0)
+	throw Xapian::InvalidArgumentError("Parameter delta is invalid.");
+    need_stat(AVERAGE_LENGTH);
+    need_stat(DOC_LENGTH);
+    need_stat(DOC_LENGTH_MIN);
+    need_stat(DOC_LENGTH_MAX);
+    need_stat(COLLECTION_SIZE);
+    need_stat(COLLECTION_FREQ);
+    need_stat(WDF);
+    need_stat(WDF_MAX);
+    need_stat(WQF);
+}
+
+PL2PlusWeight *
+PL2PlusWeight::clone() const
+{
+    return new PL2PlusWeight(param_c, param_delta);
+}
+
+void
+PL2PlusWeight::init(double)
+{
+    if (get_wdf_upper_bound() == 0) {
+	// The "extra" weight object is cloned, init() called and then
+	// get_maxextra() is called and we discover that we don't need it.
+	// So we need to handle that case (which will give us 0 from
+	// get_wdf_upper_bound() here).
+	lower_bound = upper_bound = 0;
+	return;
+    }
+
+    cl = param_c * get_average_length();
+
+    double base_change(1.0 / log(2.0));
+    double mean = double(get_collection_freq()) / get_collection_size();
+    P1 = mean * base_change + 0.5 * log2(2.0 * M_PI);
+    P2 = log2(mean) + base_change;
+
+    double wdfn_lower = log2(1 + cl / get_doclength_upper_bound());
+    double wdfn_upper =
+	get_wdf_upper_bound() * log2(1 + cl / get_doclength_lower_bound());
+
+    // Calculate the lower bound on the weight.
+    double P_min =
+	P1 + (wdfn_lower + 0.5) * log2(wdfn_lower) - P2 * wdfn_lower;
+    lower_bound = get_wqf() * P_min / (wdfn_upper + 1.0);
+
+    // Calculate the upper bound on the weight.
+    double P_max =
+	P1 + (wdfn_upper + 0.5) * log2(wdfn_upper) - P2 * wdfn_upper;
+    upper_bound = get_wqf() * P_max / (wdfn_lower + 1.0);
+
+    upper_bound -= lower_bound;
```

Comment:
The calculation of `upper_bound` is preserved unchanged from `pl2weight.cc`, which is incorrect as `get_maxpart()` is just `return upper_bound;` and we've changed the formula used in `get_sumpart()` - `get_maxpart()` is meant to return an upper bound on what `get_sumpart()` can return, and it's important that it's as tight a bound as we can provide.  Similarly, the calculation of `lower_bound` is unchanged.

The calculation of `lower_bound` and `upper_bound` needs to reflect the modifications made to the formula between PL2 and PL2+ - essentially the changes you make to `get_sumpart()` need to be reflected in the calculations here.